service/history/task_executor.go

```diff
@@ -0,0 +1,210 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/queues"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/service/history/workflow"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+)
+
+// CallbackAddressingTask is a task that addresses a workflow callback.
+type CallbackAddressingTask interface {
+	tasks.Task
+	GetCallbackID() string
+	GetTransitionCount() int32
+}
+
+func getWorkflowExecutionContextForTask(
+	ctx context.Context,
+	shardContext shard.Context,
+	workflowCache wcache.Cache,
+	task tasks.Task,
+) (workflow.Context, wcache.ReleaseCacheFunc, error) {
+	return getWorkflowExecutionContext(
+		ctx,
+		shardContext,
+		workflowCache,
+		namespace.ID(task.GetNamespaceID()),
+		&commonpb.WorkflowExecution{
+			WorkflowId: task.GetWorkflowID(),
+			RunId:      task.GetRunID(),
+		},
+	)
+}
+
+func getWorkflowExecutionContext(
+	ctx context.Context,
+	shardContext shard.Context,
+	workflowCache wcache.Cache,
+	namespaceID namespace.ID,
+	execution *commonpb.WorkflowExecution,
+) (workflow.Context, wcache.ReleaseCacheFunc, error) {
+	// workflowCache will automatically use short context timeout when
+	// locking workflow for all background calls, we don't need a separate context here
+	weContext, release, err := workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		shardContext,
+		namespaceID,
+		execution,
+		workflow.LockPriorityLow,
+	)
+	if common.IsContextDeadlineExceededErr(err) {
+		err = consts.ErrResourceExhaustedBusyWorkflow
+	}
```

Comment:
I think this can conversion be removed now: https://github.com/temporalio/temporal/pull/4341/files#diff-53066639dfa160c9360f0e33b29a37460691840bd9d86413be8bc839faa49623R240

This method and the one above are moved from another file, so I guess comments can be addressed in a separate PR?

Reply:
Yeah, I just copied the code as is without modifying the comments.
We should do all of the refactoring and better commenting in a separate PR IMHO.